client/landing/stepper/declarative-flow/internals/steps-repository/free-setup/index.tsx

```diff
@@ -29,9 +39,109 @@ const FreeSetup: Step = function FreeSetup( { navigation } ) {
 	const [ base64Image, setBase64Image ] = React.useState< string | null >();
 	const [ siteTitle, setComponentSiteTitle ] = React.useState( '' );
 	const [ tagline, setTagline ] = React.useState( '' );
+	const [ generatedDomainSuggestion, setGeneratedDomainSuggestion ] =
+		React.useState< DomainSuggestion | null >();
+	const [ generatedTopLevelDomain, setGeneratedTopLevelDomain ] = React.useState< string >( '' );
+	const [ generatedSiteName, setGeneratedSiteName ] = React.useState< string >( '' );
+	const [ showGeneratedDomainFormField, setShowGeneratedDomainFormField ] =
+		React.useState< boolean >( false );
 	const { setSiteTitle, setSiteDescription, setSiteLogo } = useDispatch( ONBOARD_STORE );
 	const state = useSelect( ( select ) => select( ONBOARD_STORE ) ).getState();
 
+	const generateNewRandomDomain = async ( siteTitle: string ) => {
+		await wpcom.req
+			.get(
+				{
+					path: '/domains/suggestions',
+					apiNamespace: 'rest/v1.1',
+				},
+				{
+					managed_subdomains: 'wordpress.com',
+					managed_subdomain_options:
+						'exact_match,random_name,random_name_format=adjective,suffix=query',
+					managed_subdomain_quantity: 1,
+					quantity: 1,
+					http_envelope: 1,
+					query: siteTitle,
+				}
+			)
+			.then( ( result: DomainSuggestion[] ) => {
+				if ( result[ 0 ] ) {
+					setGeneratedDomainSuggestion( result[ 0 ] );
+					const domainCartItem = domainRegistration( {
+						domain: result[ 0 ].domain_name,
+						productSlug: '',
+					} );
+					setDomainCartItem( domainCartItem );
+					setShowGeneratedDomainFormField( true );
+				} else {
+					setGeneratedDomainSuggestion( null );
+					setDomainCartItem( undefined );
+				}
+			} )
+			.catch( () => {
+				// TODO: add error message
+				setGeneratedDomainSuggestion( null );
+				setDomainCartItem( undefined );
+			} );
+	};
+
+	const debounceGenerateRandomDomainAPI = useMemo(
+		() => debounce( generateNewRandomDomain, 1000 ),
+		[]
+	);
+
+	const handleSubmit = async ( event: FormEvent ) => {
+		event.preventDefault();
+		setInvalidSiteTitle( ! siteTitle.trim().length );
+
+		setSiteDescription( tagline );
+		setSiteTitle( siteTitle );
+
+		if ( selectedFile && base64Image ) {
+			try {
+				setSiteLogo( base64Image );
+			} catch ( _error ) {
+				// TODO: communicate the error to the user
+			}
+		}
+
+		if ( siteTitle.trim().length && generatedDomainSuggestion?.domain_name ) {
+			submit?.( { siteTitle, tagline, siteSlug: generatedDomainSuggestion.domain_name } );
+		}
+	};
+
+	const domainGeneratorFormField = () => {
+		if ( ! showGeneratedDomainFormField ) {
+			return undefined;
+		}
+		return (
+			<FormFieldset className="setup-form-field-set-domain-name">
```

Comment:
`NB`: I think this might not be a11y friendly since screen readers will tab and skip this entirely. Maybe we should consider adding tabIndex=0 to the element? Not sure if there are better ways to handle this.

Reply:
Thanks for pointing this out. I'm noting this down, so I can circle back.